### PR TITLE
Turn projections into copies in CopyProp.

### PR DESCRIPTION
**Pull request by cjgillot**
The current implementation can leave behind projections that are moved out several times.

This PR widens the check to turn such moves into copies: a move out of a projection of a copy is equivalent to a copy of the original projection.

### CODE REVIEW

**Comment by compiler-errors** [comment]
not my place of expertise, r? compiler

**Comment by michaelwoerister** [comment]
r? mir-opt

**Comment by oli-obk** [comment]
> a move out of a projection of a copy is equivalent to a copy of the original projection.

Please leave that as a comment on the `!deref` check

r=me

File: compiler/rustc_mir_transform/src/copy_prop.rs [hunk 1]
```diff
@@ -153,8 +153,8 @@ impl<'tcx> MutVisitor<'tcx> for Replacer<'_, 'tcx> {
 
     fn visit_operand(&mut self, operand: &mut Operand<'tcx>, loc: Location) {
         if let Operand::Move(place) = *operand
-            && let Some(local) = place.as_local()
-            && !self.fully_moved.contains(local)
+            && !place.has_deref()
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/copy_prop.rs:156]:
```suggestion
            // A move out of a projection of a copy is equivalent to a copy of the original projection.
            && !place.has_deref()
```

**Pull request merged**